### PR TITLE
Add --user flag to pip.installed state

### PR DESCRIPTION
**Pull request by xeacott**
### What does this PR do?
Adds feature for pip.installed to support --user argument.

### What issues does this PR fix or reference?
 #51847 

### Previous Behavior
Currently, pip.installed state doesn't support it in any release.

### New Behavior
Now, pip.installed state does support this behavior.

### Tests written?

No (marked as WIP as setting environment up to tests to run)

### Commits signed with GPG?

Yes

### CODE REVIEW

**Comment by Oloremo** [comment]
Any updates on this? I'm willing to test it if it's needed.

**Comment by waynew** [comment]
@xeacott there are a couple of failures here... and some merge problems 😝

**Comment by xeacott** [comment]
Yes, test if you'd like like! I apologize for the delay on this but working on getting Docker setup to test for the various OS's has proved to be a slight learning curve, nonetheless I still believe the logic implemented should be fine for these tests

File: tests/integration/states/test_pip_state.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ def test_pip_installed_user_install_true(self):
         self.run_state('pip.installed', user_install=True, name=pkg)
 
         # Check if the path exists instead of adding to sys.paths
-        path = (r"{}/{}".format(USER_SITE, pkg))
+        path = ('{}/{}'.format(USER_SITE, pkg))
```
Comment (twangboy) [inline tests/integration/states/test_pip_state.py]: This may not work in Windows... maybe use...
```
os.path.join(USER_SITE, pkg)
```

Comment (twangboy) [reply tests/integration/states/test_pip_state.py]: Unless that's some other type of path...

File: tests/integration/states/test_pip_state.py [hunk 2]
```diff
@@ -110,16 +110,18 @@ def test_pip_installed_user_install_true(self):
...
+        site_pkg = os.path.join(venv_dir, 'lib/site-packages')
 
         with VirtualEnv(self, venv_dir):
             pkg = 'nose'
             self.run_state('pip.removed', name=pkg, bin_env=venv_dir)
             self.run_state('pip.installed', user_install=False, name=pkg, bin_env=venv_dir)
 
             # Check if the path exists instead of adding to sys.paths
-            path = (r"{}/{}".format(site_pkg, pkg))
+            path = ('{}/{}'.format(site_pkg, pkg))
```
Comment (twangboy) [inline tests/integration/states/test_pip_state.py]: Same here...

File: tests/integration/states/test_pip_state.py [hunk 3]
```diff
@@ -110,16 +110,18 @@ def test_pip_installed_user_install_true(self):
         self.assertEqual(location, USER_SITE)
 
     def test_pip_installed_user_install_false(self):
-        venv_dir = os.path.join(RUNTIME_VARS.TMP, r'pip_user_install')
-        site_pkg = os.path.join(venv_dir, r'lib/site-packages')
+        venv_dir = os.path.join(
+            RUNTIME_VARS.TMP, 'pip_installed_user_install_false'
+        )
+        site_pkg = os.path.join(venv_dir, 'lib/site-packages')
```
Comment (twangboy) [inline tests/integration/states/test_pip_state.py]: perhaps...
```
os.path.join(venv_dir, 'lib', 'site-packages')
```

**Review by twangboy — CHANGES_REQUESTED** [review]

**Comment by Oloremo** [comment]
@xeacott any plans to resolve conflicts and bring this PR up to date?

**Comment by xeacott** [comment]
> @xeacott any plans to resolve conflicts and bring this PR up to date?

Yep, will be cleaning this up very shortly. Had some downtime just getting up again.

File: salt/modules/pip.py [hunk 4]
```diff
@@ -1154,6 +1171,8 @@ def list_(prefix=None,
           user=None,
           cwd=None,
           env_vars=None,
+          user_install=False,
+          verbose=False,
```
Comment (twangboy) [inline salt/modules/pip.py]: Where did this come from? Part of the merge conflict resolution maybe? Need to figure out where the docs forthis went or add them if you added this.

File: salt/modules/pip.py [hunk 5]
```diff
@@ -1220,6 +1246,84 @@ def list_(prefix=None,
     return packages
 
 
+def list_json(bin_env=None,
```
Comment (twangboy) [inline salt/modules/pip.py]: You mentioned getting rid of this... is that no longer the case?

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by max-arnold** [comment]
@waynew I believe there is no need to backport this PR. Instead, it is better to port more generic one that allows arbitrary `pip` arguments: https://github.com/saltstack/salt/pull/52327

**Pull request merged**